src/main/java/shop/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+import shop.repository.CartItemRepository;
+import shop.repository.ShoppingCartRepository;
+import shop.service.ShoppingCartService;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public void createShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCartById(Authentication authentication) {
+        Long userId = getUserIdFromAuthentication(authentication);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found by user id: " + userId));
+        Set<CartItemResponseDto> cartItemResponseDtos = cartItemMapper
+                .toCartItemResponseDtos(shoppingCart.getCartItems());
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(cartItemResponseDtos);
+        return shoppingCartResponseDto;
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            Authentication authentication, AddBookToCartRequestDto requestDto) {
+        Long userId = getUserIdFromAuthentication(authentication);
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can`t get shopping cart for user with id : " + userId));
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+        CartItem existingCartItem = cartItems.stream()
+                .filter(cartItem -> cartItem.getBook().getId().equals(requestDto.getBookId()))
+                .findFirst()
+                .orElse(null);
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(requestDto.getQuantity() + existingCartItem.getQuantity());
+        } else {
+            CartItem cartItem = new CartItem();
+            cartItem.setShoppingCart(shoppingCart);
+            cartItem.setQuantity(requestDto.getQuantity());
+            cartItem.setBook(bookRepository.findById(requestDto.getBookId())
+                    .orElseThrow(() -> new EntityNotFoundException("Can`t get book by id: "
+                            + requestDto.getBookId())));
+            cartItems.add(cartItem);
+            cartItemRepository.save(cartItem);
+        }
+        shoppingCart.setCartItems(cartItems);
```

Comment:
Since you have separate methods for adding a book to the cart and updating the number of books, you don't need to write such complex logic. Unless you can check whether there is no such book in the cart and throw out an error that such a book has already been added to the cart (do not forget to return the corresponding status code to the user)